lib/filterx/object-datetime.c

```diff
@@ -200,6 +201,83 @@ _repr(FilterXObject *s, GString *repr)
   return TRUE;
 }
 
+const gchar *
+_strptime_get_time_str_from_object(FilterXObject *obj)
+{
+  if (!obj)
+    return NULL;
+
+  if (filterx_object_is_type(obj, &FILTERX_TYPE_NAME(string)))
+    return filterx_string_get_value(obj, NULL);
+
+  if (filterx_object_is_type(obj, &FILTERX_TYPE_NAME(message_value)))
+    {
+      gsize len;
+      if (filterx_message_value_get_type(obj) == LM_VT_STRING)
+        return filterx_message_value_get_value(obj, &len);
```

Comment:
`len` is igonred here, which can cause problems in some cases.

In a `LogMessage`, we can have "indirect" values, which point to a "direct" value's slice with an offset and len. In this case, the returned `const gchar *` is not NUL terminated, but its `len` is returned. We need to take care of `len` in all the call sites, or the processing function will look for the closing NUL.

One recurring way to handle this is to use the `APPEND_ZERO` macro: https://github.com/syslog-ng/syslog-ng/blob/master/lib/str-utils.h#L64

I think we should use it here, too.

---

An example config which simulates this problem, though it is far from anything being used in real life:
```
@version: current

log {
  source { example-msg-generator(num(1) template("111111")); };
  parser { regexp-parser(patterns("(?<ms_str>1).*")); };

  filterx {
    $datetime = strptime($ms_str, "%f");
  };

  destination { stdout(template("ms_str: $ms_str datetime: $datetime\n")); };
};
```
actual output:
```
ms_str: 1 datetime: -2243466061.111111+01:00
```
expected output:
```
ms_str: 1 datetime: -2243466061.100000+01:00
```